planner/core/integration_test.go

```diff
@@ -3134,3 +3134,14 @@ func (s *testIntegrationSuite) TestMultiColMaxOneRow(c *C) {
 		tk.MustQuery("explain format = 'brief' " + tt).Check(testkit.Rows(output[i].Plan...))
 	}
 }
+
+func (s *testIntegrationSuite) TestIssue23736(c *C) {
+	tk := testkit.NewTestKit(c, s.store)
+	tk.MustExec("use test")
+	tk.MustExec("create table t0(a int, b int, c int as (a + b) virtual, unique index (c) invisible);")
+	tk.MustExec("create table t1(a int, b int, c int as (a + b) virtual);")
+	tk.MustExec("insert into t0(a, b) values (12, -1), (8, 7);")
+	tk.MustExec("insert into t1(a, b) values (12, -1), (8, 7);")
+	tk.MustQuery("select /*+ stream_agg() */ count(1) from t0 where c > 10 and b < 2;").Check(testkit.Rows("1"))
```

Comment:
the result is wrong:
``` sql
mysql> select * from t0;
+------+------+------+
| a    | b    | c    |
+------+------+------+
|    5 |    1 |    6 |
+------+------+------+
1 row in set (0.00 sec)

mysql> select /*+ stream_agg() */ count(1) from t0 where c > 10 and b < 2;
+----------+
| count(1) |
+----------+
|        1 |
+----------+
1 row in set (0.00 sec)

mysql> desc select /*+ stream_agg() */ count(1) from t0 where c > 10 and b < 2;
+--------------------------------+---------+-----------+---------------+--------------------------------+
| id                             | estRows | task      | access object | operator info                  |
+--------------------------------+---------+-----------+---------------+--------------------------------+
| StreamAgg_9                    | 1.00    | root      |               | funcs:count(1)->Column#5       |
| └─Selection_18                 | 1.00    | root      |               | gt(test.t0.c, 10)              |
|   └─Projection_17              | 1.00    | root      |               | test.t0.b, test.t0.c           |
|     └─TableReader_16           | 0.11    | root      |               | data:Selection_15              |
|       └─Selection_15           | 0.11    | cop[tikv] |               | lt(test.t0.b, 2)               |
|         └─TableFullScan_14     | 1.00    | cop[tikv] | table:t0      | keep order:false, stats:pseudo |
+--------------------------------+---------+-----------+---------------+--------------------------------+
6 rows in set (0.00 sec)
```

Reply:
Nice catch!

An example wrong query plan:
```
+------------------------------+---------+-----------+---------------+------------------------------------+
| id                           | estRows | task      | access object | operator info                      |
+------------------------------+---------+-----------+---------------+------------------------------------+
| HashAgg_13                   | 1.00    | root      |               | funcs:count(Column#6)->Column#5    |
| └─TableReader_14             | 1.00    | root      |               | data:HashAgg_6                     |
|   └─HashAgg_6                | 1.00    | cop[tikv] |               | funcs:count(1)->Column#6           |
|     └─Selection_12           | 0.11    | cop[tikv] |               | lt(test.t0.b, 2), lt(test.t0.b, 2) |
|       └─TableFullScan_11     | 1.00    | cop[tikv] | table:t0      | keep order:false, stats:pseudo     |
+------------------------------+---------+-----------+---------------+------------------------------------+
```

The reason is that in `getOriginalPhysicalTableScan()`, the PhysicalTableScan.filterCondition and path.TableFilters share the same underlying array.

https://github.com/pingcap/tidb/blob/91bcb21b57bd07f77987e144dc63f1e8efc793ce/planner/core/find_best_task.go#L1744-L1757

When it is modified in addPushedDownSelection() => SplitSelCondsWithVirtualColumn()
https://github.com/pingcap/tidb/blob/cd5b91fa0388091d720e8554161b983ee980eab0/planner/core/find_best_task.go#L1527

https://github.com/pingcap/tidb/blob/91bcb21b57bd07f77987e144dc63f1e8efc793ce/planner/core/find_best_task.go#L1728-L1729

https://github.com/pingcap/tidb/blob/91bcb21b57bd07f77987e144dc63f1e8efc793ce/planner/core/find_best_task.go#L1184-L1193

Actually, what L1189 doing is changing the underlying array from [c > 10, b < 2] into [b < 2, b < 2]

The original AccessPath.TableFilters is also modified, all the plans generated after this are all wrong.